File: canonical.go
package pact

import (
	"bytes"
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
)

// canonicalJSON serializes a value to RFC 8785 JSON Canonicalization Scheme (JCS).
//
// JCS guarantees deterministic output:
//   - Object keys sorted lexicographically (by Unicode code points)
//   - No whitespace
//   - Numbers serialized per ES2015 Number.toString()
//   - Strings use minimal escaping
//   - No trailing commas
//
// This is essential for signing: the same logical value must always produce
// the same byte sequence, regardless of which implementation serializes it.
func canonicalJSON(v interface{}) ([]byte, error) {
	// First, round-trip through encoding/json to get a normalized representation.
	// This handles struct tags, omitempty, etc.
	raw, err := json.Marshal(v)
	if err != nil {
		return nil, fmt.Errorf("pact: canonical JSON marshal failed: %w", err)
	}

	// Parse into generic interface{} for canonical re-serialization
	var parsed interface{}
	decoder := json.NewDecoder(bytes.NewReader(raw))
	decoder.UseNumber() // Preserve number precision
	if err := decoder.Decode(&parsed); err != nil {
		return nil, fmt.Errorf("pact: canonical JSON decode failed: %w", err)
	}

	var buf bytes.Buffer
	if err := writeCanonical(&buf, parsed); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// writeCanonical writes a value in JCS canonical form.
func writeCanonical(buf *bytes.Buffer, v interface{}) error {
	switch val := v.(type) {
	case nil:
		buf.WriteString("null")

	case bool:
		if val {
			buf.WriteString("true")
		} else {
			buf.WriteString("false")
		}

	case json.Number:
		// RFC 8785: numbers must be serialized per ES2015 Number.toString()
		// For integers, this is just the decimal representation.
		// For floats, we need to handle edge cases.
		// Our protocol uses no floats (timestamps are strings, amounts are integers in cents).
		// But we handle both for correctness.
		if i, err := val.Int64(); err == nil {
			buf.WriteString(strconv.FormatInt(i, 10))
		} else if f, err := val.Float64(); err == nil {
			buf.WriteString(canonicalFloat(f))
		} else {
			return fmt.Errorf("pact: invalid number in canonical JSON: %s", val.String())
		}

	case string:
		// RFC 8785: strings use standard JSON escaping (RFC 8259)
		// encoding/json.Marshal handles this correctly for strings
		escaped, err := json.Marshal(val)
		if err != nil {
			return fmt.Errorf("pact: failed to escape string: %w", err)
		}
		buf.Write(escaped)

	case map[string]interface{}:
		buf.WriteByte('{')

		// RFC 8785: keys sorted by Unicode code point order
		keys := make([]string, 0, len(val))
		for k := range val {
			keys = append(keys, k)
		}
		sort.Strings(keys)

		for i, k := range keys {
			if i > 0 {
				buf.WriteByte(',')
			}
			// Write key
			keyBytes, err := json.Marshal(k)
			if err != nil {
				return fmt.Errorf("pact: failed to escape key: %w", err)
			}
			buf.Write(keyBytes)
			buf.WriteByte(':')
			// Write value
			if err := writeCanonical(buf, val[k]); err != nil {
				return err
			}
		}

		buf.WriteByte('}')

	case []interface{}:
		buf.WriteByte('[')
		for i, item := range val {
			if i > 0 {
				buf.WriteByte(',')
			}
			if err := writeCanonical(buf, item); err != nil {
				return err
			}
		}
		buf.WriteByte(']')

	default:
		return fmt.Errorf("pact: unsupported type in canonical JSON: %T", v)
	}

	return nil
}

// canonicalFloat serializes a float64 per ES2015 Number.toString().
// This handles the edge cases required by RFC 8785.
func canonicalFloat(f float64) string {
	// Use strconv with the shortest representation that round-trips
	return strconv.FormatFloat(f, 'G', -1, 64)
}

File: sdk/python/pact/canonical.py
"""RFC 8785 JSON Canonicalization Scheme (JCS).

Produces deterministic JSON output for signing:
- Object keys sorted lexicographically by Unicode code points
- No whitespace between tokens
- Numbers per ECMAScript 2015 Number.prototype.toString()
- Strings use standard JSON escaping (RFC 8259)
"""

import json
import math
from typing import Any


def canonical_json(obj: Any) -> bytes:
    """Serialize a Python object to canonical JSON bytes (RFC 8785 JCS).

    The input can be any JSON-serializable Python value: dict, list, str,
    int, float, bool, or None.

    Returns UTF-8 encoded bytes that are deterministic — the same logical
    value always produces the same byte sequence across implementations.
    """
    # If the input is a dataclass or has a to_dict method, convert first
    if hasattr(obj, "to_dict"):
        obj = obj.to_dict()

    # Round-trip through json to normalize (handles dataclasses, enums, etc.)
    normalized = json.loads(json.dumps(obj, default=_json_default))
    return _write_canonical(normalized).encode("utf-8")


def _json_default(obj: Any) -> Any:
    """Default serializer for objects json.dumps can't handle."""
    if hasattr(obj, "to_dict"):
        return obj.to_dict()
    raise TypeError(f"Object of type {type(obj)} is not JSON serializable")


def _write_canonical(val: Any) -> str:
    """Write a value in JCS canonical form."""
    if val is None:
        return "null"
    if isinstance(val, bool):
        return "true" if val else "false"
    if isinstance(val, int):
        return str(val)
    if isinstance(val, float):
        return _canonical_float(val)
    if isinstance(val, str):
        return _canonical_string(val)
    if isinstance(val, dict):
        # RFC 8785: keys sorted by Unicode code point order
        parts = []
        for key in sorted(val.keys()):
            k = _canonical_string(key)
            v = _write_canonical(val[key])
            parts.append(f"{k}:{v}")
        return "{" + ",".join(parts) + "}"
    if isinstance(val, list):
        parts = [_write_canonical(item) for item in val]
        return "[" + ",".join(parts) + "]"
    raise TypeError(f"Unsupported type in canonical JSON: {type(val)}")


def _canonical_string(s: str) -> str:
    """Serialize a string with standard JSON escaping (RFC 8259).

    Uses Python's json.dumps which handles all required escaping.
    """
    return json.dumps(s, ensure_ascii=False)


def _canonical_float(f: float) -> str:
    """Serialize a float per ECMAScript 2015 Number.prototype.toString().

    Handles special cases required by RFC 8785.
    """
    if math.isnan(f) or math.isinf(f):
        raise ValueError(f"Cannot serialize {f} in canonical JSON")
    if f == 0.0:
        return "0"
    # Use repr to get shortest representation that round-trips
    s = repr(f)
    # Python's repr already gives shortest roundtrip representation
    # but we need to match ES2015 formatting
    if "e" in s or "E" in s:
        return s
    # Remove trailing zeros after decimal point
    if "." in s:
        s = s.rstrip("0").rstrip(".")
    return s
